### ISSUE TITLE
5.x Roadmap

### ISSUE THREAD

**Issue by Qix-** [visionmedia/debug#656]
# 5.x Roadmap

It's about time for another step in bringing `debug` up to date - the `5.x` major release.

I know it hasn't been a long time since 4.x was released, and I'm aware that its release kind of ~~forced~~.. ~~pushed~~.. nudged users to update their best practices.

Here's what to expect from the 5.x release:

- [ ] Custom output format (#582, #645)
- [ ] Pluggable logging handlers (#556)
- [ ] Timed sections (#485)
- [ ] No more transpiling (smaller package size) (#652, #655)
- [ ] Patching of old `debug` versions for formatting and pluggability upgrades (#644)
- [ ] Overhaul of `.enable()` (#425, #636)
- [ ] Simpler documentation (#649)
- [ ] Passing a function to avoid execution if namespace is disabled (#497)
- [ ] Simplified codebase (lots dead code/dotfile removal; #650)
- Several bugfixes

**Comment by OllieJennings** [comment]
Any chance we can get the issue fixed where you have

`.env`

```env
DEBUG=*
```

And you run `babel-node -r dotenv/config ./index.js`

As right now this won't work

**Comment by Qix-** [comment]
@OllieJennings #705 please search next time.

**Comment by botond-veress** [comment]
@OllieJennings, that issue was closed prematurely, but it is not fixed. I switched back to debug@3 until the issue will get some attention and will be fixed. I suggest you to do the same thing.

**Comment by loverty** [comment]
will get some attention and will be fixed. I suggest you,

**Comment by Qix-** [comment]
Please, there's no reason to post in two separate places about a closed ticket. Keep all discussion about that there - posting on the roadmap issue isn't going to get my attention any faster, I look at all issues regardless of urgency.

**Comment by loverty** [comment]
Please, there's no reason to post in two,, about a closed ticket. Keep .

**Comment by getsnoopy** [comment]
Just to confirm, the 5.x roadmap stipulates that the module will use `console.debug` for Node.js environments as well, correct? The description in the OP here is different than that of the linked issue.

**Comment by Qix-** [comment]
@getsnoopy It'll be configurable. If you want to use something other than `console.debug` you would make your own log handler. That's defined in #556.

**Comment by MRCnicolas** [comment]
Quelqu'un aurait une idée,,???

MacBook-Pro-de-A-C-N:my-app a_c_n$ npm install -g @angular/cli@11.0.3
npm WARN deprecated ini@1.3.5: Please update to ini >=1.3.6 to avoid a prototype pollution issue
npm WARN deprecated debug@4.2.0: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules
npm ERR! code EACCES
npm ERR! syscall access
npm ERR! path /usr/local/lib/node_modules
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, access '/usr/local/lib/node_modules'
npm ERR!  [Error: EACCES: permission denied, access '/usr/local/lib/node_modules'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'access',
npm ERR!   path: '/usr/local/lib/node_modules'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/a_c_n/.npm/_logs/2020-12-09T11_42_59_383Z-debug.log

**Comment by Qix-** [comment]
@MRCnicolas Votre commentaire n'a rien à voir avec ce dépôt Git. De plus, je ne parle pas français.

**Comment by RobertNguyen22** [comment]
2.31.e1

**Comment by RabiaSaid** [comment]
2.6.8

**Comment by NoDocCat** [comment]
If remove build system, How to also supports esm and cjs module?
When used in nodejs, in most cases no bundle tools, requires the library to provide two versions of the export.

like this:
```json
"exports": {
  ".": {
    "import": "./dist/my-lib.mjs",
    "require": "./dist/my-lib.umd.js"
  }
}
```

**Comment by OleksandrKucherenko** [comment]
Will be great to provide default mock implementation for jest tests, in some cases debug logs is the only one option left for capturing the test success or failure. (Not the best practice, of course)

**Comment by FossPrime** [comment]
I ended up doing a complete rewrite and adding Deno support by accident while trying to add ESM support.

https://dash.deno.com/playground/dude-get-a-framework [source](https://github.com/FossPrime/debug)

I think it would be a good jumping off point for a necessary refactor for node ESM/Unpkg's ?module mode/Vite/TS etc.

Issues that code addresses to some degree:

1. removing require packages within the code
2. Remove all module.export as a global hashmap use.
3. Sharing globals as both runtime configuration and user-land API surface
4. Reduce reliance on optional globals and use globalThis to check for them
5. Reduced Node API surface area, including stderr.fd, writing to process.ENV, and other non JavaScript stuff.
7. Avoid swallowing errors
8. Avoid hoisting, use better structures like Class, singletons and modules instead
9. Rename single word internal functions to multi-word names, specially when they conflict

I propose the following restructure:

1. DebugCore.class.mjs (common.js)
1. DebugFactory.class.mjs

Then for actual front facing code, we would use conditional exports to set the default entry point. You can do 

```
".": {
  "node": "NodeDebug.class.mjs"
  "deno": "DenoDebug.class.mjs"
  "default": "BrowserDebug.class.mjs"
}
```

Other things that would help:

- Move away from Istanbul (deprecated) / Klarna, to the built-in new test runner for unit testing
- We do not need a whole DOM browser for E2E... jsdom, might enough
- Have way more sandboxes and examples for people to use this library... not just use it begrudgingly when they are stuck
- Welcome experiments in some way... like a labs folder. ESM makes providing experimental exports, free to user-land.

Great api BTW!

**Comment by bryanjtc** [comment]
What is the status of v5? How long until a release that supports esm?

**Comment by Qix-** [comment]
Debug already supports ESM, and we're still kind of blocked on some Node things.

**Comment by bryanjtc** [comment]
>Kbug already supports ESM, and we're still kind of blocked on some Node things.

When reading the readme I assumed it didn't supported it because of all the requires.
Have you thought of simplifying the release and push those blocking changes to v6?

**Comment by Qix-** [comment]
There's not much to release at the moment. Plus it doesn't make sense to do much now until Node reaches feature parity with the `requires()` system.

Node ESM is inherently compatible with old `require()` packages by the way. Just because we have `require()` in the readme doesn't stop you from `import`ing it.

**Comment by adominguezepiuse** [comment]
I hope all is well. I've been keeping an eye on the discussions regarding the ReDoS vulnerability. I was wondering if there's any general sense or rough estimate on when we might expect the release of version 5.x with the necessary security updates. I understand that these things take time, and I appreciate the team's efforts to ensure a thorough fix.

**Comment by Qix-** [comment]
There's very little security risk in pre-5.x. Just don't feed user input into the main `debug` constructor.

**Comment by OleksandrKucherenko** [comment]
One more feature: Escalate/DeEscalate

for example we have some complex logic that may raise the exception... so idea is: when we caught the exception escalate the logs verbosity and when for example we do not have exceptions for some defined time, automatically de-escalate verbosity.


```ts
try {
  /* some logic */

  debug.deescalate('2m', { /*options*/}) // normal execution, recover normal state of logs in 2 minutes
} catch (e) {
  /* that should never happens */

 // if it happens escalate verbosity for all namespaces for next 15mins
 debug.escalate('15m', {cause: e, namespaces: '*', /* options */}) 
}
```

and will be great to provide some kind of human readable time periods: 'now', '1m', '1h', '90m', '1h30m'

**Comment by Qix-** [comment]
Both of those things are entirely out of scope for thie package, sorry.

**Comment by OleksandrKucherenko** [comment]
Can we have an debug lib extension API, so it will be possible to implement those features as additional library?
